android/modules/database/src/java/ti/modules/titanium/database/TiDatabaseProxy.java

```diff
@@ -18,145 +19,244 @@
+					parameters.add(TiConvert.toString(parameter));
 				}
-				c = db.rawQuery(sql, selectArgs);
-				if (c != null) {
-					// Most non-SELECT statements won't actually return data, but some such as
-					// PRAGMA do. If there are no results, just return null.
-					// Thanks to brion for working through the logic, based off of commit
-					// https://github.com/brion/titanium_mobile/commit/8d3251fca69e10df6a96a2a9ae513159494d17c3
-					if (c.getColumnCount() > 0) {
-						rs = new TiResultSetProxy(c);
-						if (rs.isValidRow()) {
-							rs.next(); // Position on first row if we have data.
+
+				cursor = db.rawQuery(query, parameters.toArray(new String[0]));
+				if (cursor != null) {
+
+					// Validate and set query result.
+					if (cursor.getColumnCount() > 0) {
+						result = new TiResultSetProxy(cursor);
+						if (result.isValidRow()) {
+							result.next();
 						}
-					} else {
-						c.close();
-						c = null;
-						rs = null;
 					}
-				} else {
-					// Leaving for historical purposes, but walking through several different
-					// types of statements never hit this branch. (create, drop, select, pragma)
-					rs = new TiResultSetProxy(null); // because iPhone does it this way.
 				}
+
+				// Query does not return result, use execSQL()
 			} else {
-				Object[] newArgs = null;
-				if (sqlArgs != null) {
-					newArgs = new Object[sqlArgs.length];
-					for (int i = 0; i < sqlArgs.length; i++) {
-						if (sqlArgs[i] instanceof TiBlob) {
-							newArgs[i] = ((TiBlob) sqlArgs[i]).getBytes();
-						} else {
-							newArgs[i] = TiConvert.toString(sqlArgs[i]);
-						}
-					}
-				}
-				db.execSQL(sql, newArgs);
+				db.execSQL(query, parameterObjects);
 			}
-		} catch (SQLException e) {
-			String msg = "Error executing sql: " + e.getMessage();
-			Log.e(TAG, msg, e);
-			if (c != null) {
+		} finally {
+
+			// Cleanup result.
+			if (cursor != null) {
 				try {
-					c.close();
-				} catch (SQLException e2) {
-					// Ignore
+					cursor.close();
+				} catch (Exception e) {
+					// Ignore...
 				}
 			}
-			throw e;
 		}
 
-		return rs;
+		return result;
+	}
+
+	/**
+	 * Asynchronously execute a single SQL query.
+	 * @param callback Result callback for query execution.
+	 * @param query SQL query to execute on database.
+	 * @param parameters Parameters for `query`
+	 */
+	@Kroll.method
+	public void executeAsync(final KrollFunction callback, final String query, final Object... parameters)
+	{
+		executingQueue.set(true);
+		try {
+			queue.put(new Runnable() {
+				@Override
+				public void run()
+				{
+					final TiResultSetProxy result = execute(query, parameters);
+					callback.callAsync(getKrollObject(), new Object[] { result });
+				}
+			});
+		} catch (InterruptedException e) {
+		}
+	}
+
+	/**
+	 * Synchronously execute a multiple SQL queries.
+	 * @param queries SQL queries to execute on database.
+	 */
+	@Kroll.method
+	public Object[] executeAll(final String[] queries)
```

Comment:
We should look to something like this node sqlite3 package to consider how they've approached their API: https://github.com/mapbox/node-sqlite3/wiki/API#database

Maybe not introduce prepare/Statements, but having explicit methods to say whether you want to run something without caring about the result, or wanting the first result, or calling the callback with the batch results once vs calling for each row seems useful to me.

Note that there is also this package which does explicitly run sync intentionally for performance: https://github.com/JoshuaWise/better-sqlite3/blob/master/docs/api.md

I recall @jquick-axway mentioning usage of transactions helped performance significantly.  Would we wrap this batch of queries in a transaction for the user? Would we expect them to include their own transaction management statements in this batch?